### PR TITLE
feat: add backend requests for CPV screens

### PR DESCRIPTION
**Pull request by kathaypacific**
### Description

As the title, implement the backend calls to verify phone number and handle errors.

[Happy flow video](https://valora-app.slack.com/archives/C029Z1QMD7B/p1664298511554419).

Error initiating verification flow:

https://user-images.githubusercontent.com/20150449/192523355-e0085f97-261a-47d2-91d9-5bf05b60b090.MP4

Error in sms code input:

https://user-images.githubusercontent.com/20150449/192523247-83b9f49e-042d-487d-a339-9f65cf48f1c8.MP4

Not included in this PR:
- ability to resend the code
- storing the new phone verification status
- allowing users who already verified to skip (recover account scenario)

### Other changes

N/A

### Tested

Manually and unit tests.
 
- a phone number can be successfully verified
- an error banner should be displayed if the verification process could not be successfully initiated
- the sms code input should display an error if the code was wrong
- as usual, once the phone number is verified the user should not be able to trigger the verification flow again

### How others should test

As above

### Related issues

- Relates to RET_384

### Backwards compatibility

Yes

### CODE REVIEW

**Review by jeanregisser — COMMENTED** [review]
Awesome!

File: src/verify/hooks.ts [hunk 1]
```diff
@@ -27,3 +33,87 @@ export function useAsyncKomenciReadiness() {
...
+      setVerificationStatus(PhoneNumberVerificationStatus.SUCCESSFUL)
+    } else {
+      Logger.debug(
+        TAG,
+        'startPhoneNumberVerificationSaga received error from verify phone number service'
+      )
+      setVerificationStatus(PhoneNumberVerificationStatus.FAILED)
+      // dispatch(showError(ErrorMessages.START_PHONE_VERIFICATION_FAILURE))
+    }
+  }
```
[the remark this answers is not in the document]
Comment (kathaypacific) [reply src/verify/hooks.ts]: ah yes....i updated this :) i don't like how the useAsync hook is triggered though, i needed to put early returns in there which doesn't look great but i suppose it does work better!

File: src/verify/hooks.ts [hunk 2]
```diff
@@ -27,3 +33,87 @@ export function useAsyncKomenciReadiness() {
...
+      const result = await response.json()
+      setVerificationId(result.verificationId)
+      setVerificationStatus(PhoneNumberVerificationStatus.AWAITING_USER_INPUT)
+    } else {
+      Logger.debug(
+        TAG,
+        'startPhoneNumberVerificationSaga received error from verify phone number service'
+      )
+      setVerificationStatus(PhoneNumberVerificationStatus.FAILED)
+      // dispatch(showError(ErrorMessages.START_PHONE_VERIFICATION_FAILURE))
```
[the remark this answers is not in the document]
Comment (kathaypacific) [reply src/verify/hooks.ts]: yes great question, let me take this up with Nitya to see what we want to do and address this in a separate PR

File: src/verify/hooks.ts [hunk 3]
```diff
@@ -27,3 +36,150 @@ export function useAsyncKomenciReadiness() {
...
+      `Received error from verifySmsCode service for verificationId: ${verificationId}`,
+      error
+    )
+    setVerificationStatus(PhoneNumberVerificationStatus.FAILED)
+    setSmsCode('')
+  }
+
+  useAsync(
+    async () => {
+      if (verificationCodeRequested.current) {
```
Comment (jeanregisser) [inline src/verify/hooks.ts:78]: Why is this needed? `useAsync` should guarantee it's only called once for a given set of passed dependencies.

File: src/verify/hooks.ts [hunk 4]
```diff
@@ -27,3 +36,150 @@ export function useAsyncKomenciReadiness() {
...
+          verificationCodeRequested.current = true
+          Logger.debug(
+            `${TAG}/requestVerificationCode`,
+            'Successfully initiated phone number verification with verificationId: ',
+            data.verificationId
+          )
+        } else {
+          handleRequestVerificationCodeError(new Error(await response.text()))
+        }
+      },
```
Comment (jeanregisser) [inline src/verify/hooks.ts:123]: Same here.

File: src/verify/VerificationCodeInputScreen.test.tsx [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+
+    await act(flushMicrotasksQueue)
+
+    expect(mockFetch).toHaveBeenCalledTimes(1)
+    expect(store.getActions()).toEqual(
+      expect.arrayContaining([showError(ErrorMessages.PHONE_NUMBER_VERIFICATION_FAILURE)])
+    )
+  })
+
+  it('verifies the sms code service', async () => {
```
Comment (jeanregisser) [inline src/verify/VerificationCodeInputScreen.test.tsx:88]: Did you mean?
```suggestion
  it('verifies the sms code', async () => {
```

File: src/verify/hooks.ts [hunk 6]
```diff
@@ -27,3 +36,150 @@ export function useAsyncKomenciReadiness() {
...
 }
+
+export enum PhoneNumberVerificationStatus {
+  NONE,
+  VERIFYING,
+  SUCCESSFUL,
+  FAILED,
+}
+
+export function useVerifyPhoneNumber(phoneNumber: string, countryCode: string) {
```
Comment (jeanregisser) [inline src/verify/hooks.ts:47]:
```suggestion
export function useVerifyPhoneNumber(phoneNumber: string, countryCallingCode: string) {
```

**Review by jeanregisser — COMMENTED** [review]
Amazing work! 😍 

Mostly good, just some comments trying to make the hook a bit cleaner.

File: src/verify/hooks.ts (hunk 3 above)
Comment (kathaypacific) [reply src/verify/hooks.ts:77]: actually this is for development - the hot reload would always reload the hook and it was sending too many requests. i added a comment, do you have a better idea of how to handle this?

File: src/verify/hooks.ts [hunk 7]
```diff
@@ -27,3 +36,150 @@ export function useAsyncKomenciReadiness() {
...
+
+        if (response.ok) {
+          Logger.debug(`${TAG}/validateVerificationCode`, 'Successfully verified phone number')
+          setVerificationStatus(PhoneNumberVerificationStatus.SUCCESSFUL)
+          dispatch(setPhoneNumber(phoneNumber, countryCode))
+          // TODO store verification status in new redux variable so that the
+          // existing one can be used for background migration
+        } else {
+          handleVerifySmsError(new Error(await response.text()))
+        }
```
[the remark this answers is not in the document]
Comment (kathaypacific) [reply src/verify/hooks.ts]: yes you're right - i updated this :)

File: src/navigator/types.tsx [hunk 8]
```diff
@@ -336,7 +336,8 @@ export type StackParamList = {
     | undefined
   [Screens.VerificationCodeInputScreen]: {
     registrationStep?: { step: number; totalSteps: number }
-    e164Number?: string
+    e164Number: string
+    countryCode: string
```
[the remark this answers is not in the document]
Comment (kathaypacific) [reply src/navigator/types.tsx]: yeah good idea, i updated

File: src/verify/VerificationStartScreen.tsx [hunk 9]
```diff
@@ -143,25 +136,6 @@ function VerificationStartScreen({
...
-      cachedCountryCallingCode === countryCallingCode
-    ) {
-      return true
-    }
-
-    const { SANCTIONED_COUNTRY } = getCountryFeatures(phoneNumberInfo.countryCodeAlpha2)
-    if (SANCTIONED_COUNTRY) {
-      dispatch(showError(ErrorMessages.COUNTRY_NOT_AVAILABLE))
-      return false
-    }
```
[the remark this answers is not in the document]
Comment (kathaypacific) [reply src/verify/VerificationStartScreen.tsx:159]: i don't think so - the country picker already filters out the sanctioned countries so a user wouldn't be able to add a phone number from there right? https://github.com/valora-inc/wallet/blob/6f7dc98ab95e632abd07afcf8942d59606465e46/src/onboarding/registration/SelectCountry.tsx#L27-L33

File: src/verify/VerificationCodeInputScreen.tsx [hunk 10]
```diff
@@ -92,7 +107,9 @@ function VerificationCodeInputScreen({
           inputPlaceholder={t('phoneVerificationInput.codeInputPlaceholder')}
           onInputChange={setCode}
           shouldShowClipboard={(content) =>
-            !!content && content.length === VERIFICATION_CODE_LENGTH
+            Platform.OS === 'android' &&
```
[the remark this answers is not in the document]
Comment (kathaypacific) [reply src/verify/VerificationCodeInputScreen.tsx]: on second thought we don't really need to restrict it on android - i thought that we don't need to show the paste button on ios because we do the auto populate thing but it doesn't really hurt to have it. i removed the check

**Comment by kathaypacific** [comment]
@jeanregisser ready for review again!

File: src/verify/VerificationStartScreen.tsx (hunk 9 above)
[the remark this answers is not in the document]
Comment (jeanregisser) [reply src/verify/VerificationStartScreen.tsx:159]: Right! They shouldn't 👍

File: src/verify/hooks.ts (hunk 3 above)
Comment (jeanregisser) [reply src/verify/hooks.ts:77]: oh interesting. Not sure there's a better way then.

**Review by jeanregisser — APPROVED** [review]
🚀 🚀 🚀
